### ISSUE TITLE
Ocean Compatibility

### ISSUE THREAD

**Issue by haubna** [ValkyrienSkies/Valkyrien-Skies-2#595]
### Mod Name

Physics Mod

### This issue occurs when only Valkyrien Skies, addons, and the mod I have specified are installed and no other mods

- [X] I have tested this issue and it occurs with only Valkyrien Skies, addons, and the mod I have specified

### Minecraft Version

1.20.x

### Mod Loader

Fabric

### Issue description

Hi, creator of the physics mod here. I would like to make the mod somewhat compatible with valkyrien skies. Is there a way to fetch the entity position of your physics entities on the client side without using VSPhysicsEntity.getRenderTransform?

I am basically using entity.getX() (getY(), getZ()) for the ships currently and those just return static numbers like those: 

`-2.86658515E7 / 64.0 / 1.22900625E7`

but they should be

` 809.153109844774 / 62.93042876152322 / 191.969636304304`

If there is no other way around it I will have to stick to getRenderTransform anyways (I was just trying to avoid any dependencies since this could break when you update your mod).

### Issue reproduction

Build a ship with Eureka and put it in the ocean with ocean physics enabled from physics mod.

**Comment by NullHarp** [comment]
Hi im not a vs2 dev but just wanted to let you know ships arent entities but are their own system, more similar to a distant cousin of the entity system rather than directly the same thing, good luck with compat, would be cool

**Comment by haubna** [comment]
Well there are entities in place for rendering and sending states over the network at least. What's happening in the background is not really something I care about.

**Comment by NullHarp** [comment]
are you responding to me or to the OP, i think they mixed up phys entities and normal vs2 ships since they mention eureka assembly

**Comment by SuperCraftAlex** [comment]
also the positions should not be transformed automatically

**Comment by haubna** [comment]
Could you point me in the direction where I could fetch the positions of the ship on the client side and where the actual rendering for it happens?

**Comment by SuperCraftAlex** [comment]
if you want the world position, either use render transform or depend on vs2
or use Class.forName

**Comment by haubna** [comment]
Perfect, thanks! I am assuming the rendering of the ships happens in [Entity Renderer](https://github.com/ValkyrienSkies/Valkyrien-Skies-2/blob/1.18.x/main/common/src/main/kotlin/org/valkyrienskies/mod/client/VSPhysicsEntityRenderer.kt), right? So I will stick to that one.

**Comment by SuperCraftAlex** [comment]
no it isnt 
ships aae rendered as chunks in the level renderer

**Comment by bmartin127** [comment]
I think it does have some server side issues also what kind of compatibility are you going to add

**Comment by haubna** [comment]
I should be able to Mixin into your Mixins (I have done it before with I think Iris). I will let you know in case I need anything. Thanks for the help! :)

**Comment by SuperCraftAlex** [comment]
#596 adds an event hook that helps with that.
Example code (untested):
```kotlin
  try {
    val clazz = Class.forName("org.valkyrienskies.mod.common.hooks.VSGameEvents")
    val instance = clazz.getField("INSTANCE").get(null) as VSGameEvents
    val emitter = clazz.getMethod("getRenderShipModifiable").invoke(instance)
        as EventEmitterImpl<ModifiableShipRenderEvent>
    emitter.on { event ->
        // do stuff
    }
  }
  catch(e: Exception) {
      when (e) {
          is ClassNotFoundException,
          is NoSuchFieldException,
          is IllegalAccessException,
          is NoSuchMethodException -> {
              println("This VS2 version does not have the required event hooks.")
          }
          else -> throw e
      }
  }
```
With this, you only need VS2 in your dev env and it does not break if VS2 is not loaded at runtime / a VS2 version without this event hook is loaded

**Comment by SuperCraftAlex** [comment]
Note that this is only for rendering of the chunks of the ship.
Not entities on the ship or block hitboxes or anything

**Comment by haubna** [comment]
Thanks, figured that out already however I am using those 2 currently (not sure if there is a big difference, have only been testing if it works)
```Java
VSGameEvents.INSTANCE.getRenderShip().on((shipRenderEvent) -> {
	PoseStack pose = shipRenderEvent.getPoseStack();
	pose.pushPose();
	pose.translate(0, 4, 0);
});

VSGameEvents.INSTANCE.getPostRenderShip().on((shipRenderEvent) -> {
	shipRenderEvent.getPoseStack().popPose();
});
```

With this approach however the steering wheel from Eureka doesn't get transformed.

**Comment by haubna** [comment]
[This](https://github.com/ValkyrienSkies/Valkyrien-Skies-2/blob/0b7cf8a2a95f03ce2a9c22d31b410bb7ceb29745/common/src/main/java/org/valkyrienskies/mod/mixin/mod_compat/sodium/MixinSodiumWorldRenderer.java) seems to be the right one for Sodium but I haven't found the one for vanilla yet.

**Comment by SuperCraftAlex** [comment]
its MixinLevelRendererVanilla

**Comment by haubna** [comment]
Are you sure? I haven't found the part where it transforms Block Entities in there.

**Comment by walksanatora** [comment]
the thing everyone is missing: physics mod is a *client side mod*
the waves are entirely client sided where as the ships are mostly server-sided... so unlless you want to sync that this is a bad idea imo

**Comment by SuperCraftAlex** [comment]
Thats why he is doing a client side render transform

**Comment by SuperCraftAlex** [comment]
found it
its VSEntityHandler#applyRenderTransform

**Comment by haubna** [comment]
Are you sure that this is for Block Entities / Tile Entities?

**Comment by SuperCraftAlex** [comment]
It is for normal entities
not sure about tile entities

**Comment by walksanatora** [comment]
> Thats why he is doing a client side render transform

what about entities standing on ships eg: a player that is on a ship yet not in shipyard
you then have to translate that

**Comment by SuperCraftAlex** [comment]
yes there is a lot of stuff to transform to make it work correctly

**Comment by haubna** [comment]
Entities standing on top will probably be another weird edge case but first I have to find the Block Entities part :D

**Comment by haubna** [comment]
Ok so I will probably stick to this right here. This allows me to not have to inject in too many other code and renders fine without having to deal with block entities and entities as well (and coming with it all the other mod compats).

```Java
@Pseudo
@Mixin(ShipObjectClient.class)
public class MixinShipObjectClient {

	@Shadow
	private ShipTransform renderTransform;

	@Inject(at = @At("TAIL"), method = "updateRenderShipTransform", remap = false)
	public final void physicsmod$applyOceanTransformation(double partialTicks, CallbackInfo info) {
		renderTransform = ShipTransformImpl.Companion.create(
				new Vector3d(renderTransform.getPositionInWorld()).add(0.0, 4.0, 0.0),
				renderTransform.getPositionInShip(),
				renderTransform.getShipToWorldRotation(),
				renderTransform.getShipToWorldScaling());
	}

}
```

**Comment by haubna** [comment]
Is it possible to know which ship an entity is standing on on the client side? I've currently tried it like this but this doesn't seem to return a ship.

```Java
LoadedShip shipMountedTo = VSGameUtilsKt.getShipObjectEntityMountedTo(entity.level(), entity);

if (entity instanceof IEntityDraggingInformationProvider draggingProvider) {
	EntityDraggingInformation info = draggingProvider.getDraggingInformation();
	info.isEntityBeingDraggedByAShip();
	info.getLastShipStoodOn();
}```

**Comment by SuperCraftAlex** [comment]
you should mixin into the impl of VSEntityHandler#applyRenderTransform

**Comment by haubna** [comment]
I might be mistaken but VSEntityHandler#applyRenderTransform never gets called (I thought it gets called when I am on a ship).

**Comment by SuperCraftAlex** [comment]
AbstractShipyardEntityHandler#applyRenderTransform gets called for all entities that are on (standing, moving...) a ship (not sure about the player)

**Comment by haubna** [comment]
I also encountered a very weird bug that was difficult to track down. Very sporadically the view just started jittering. As it turns out that happens because I alter the render transformation + the tick delta being above or equal to 0.99999 (which doesn't apply the entityShouldBeHere and causes a jitter): https://github.com/ValkyrienSkies/Valkyrien-Skies-2/blob/b3555ab2ad2f5ff44e6d42c4a203046ffd4ff797/common/src/main/java/org/valkyrienskies/mod/mixin/client/renderer/MixinGameRenderer.java#L188

In my case I fixed it with (not very pretty but there shouldn't be any noticable difference). I thought I post it here in case it comes up somewhere else.
```Java
@ModifyVariable(method = "render", at = @At("HEAD"), ordinal = 0)
private float physicsmod$fixJitter(float tick) {
	return Math.min(0.9999f, tick);
}
```
